File: 리트코드/0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> lst = Arrays.stream(nums).boxed().toList();
        dfs(result, new ArrayList<>(), lst);
        return result;
    }

    public void dfs(List<List<Integer>> result, List<Integer> prevElements,
            List<Integer> elements) {
        if (elements.isEmpty()) {
            result.add(new ArrayList<>(prevElements));
        }
        for (Integer e : elements) {
            List<Integer> nextElements = new ArrayList<>(elements);
            nextElements.remove(e);

            prevElements.add(e);
            dfs(result, prevElements, nextElements);
            prevElements.remove(e);
        }
    }
}

File: 리트코드/0046-permutations/0046-permutations.kt
class Solution {
    fun permute(nums: IntArray): List<List<Int>> {
        val result: MutableList<List<Int>> = mutableListOf()
        val lst = Arrays.stream(nums).boxed().collect(Collectors.toList())
        
        fun dfs(prevElements:MutableList<Int>, elements:MutableList<Int>) {
            if (elements.isEmpty()) {
                result.add(prevElements.stream().collect(Collectors.toList()))
                return
            }
            for (e in elements) {
                val nextElements = ArrayList(elements)
                nextElements.remove(e)

                prevElements.add(e)
                dfs(prevElements, nextElements)
                prevElements.remove(e)
            }
        }
        
        dfs(mutableListOf(),lst)

        return result
    }
}